packages/contracts-bedrock/test/scripts/L2Genesis.t.sol

```diff
@@ -0,0 +1,246 @@
+    DeployConfig public constant cfg =
+        DeployConfig(address(uint160(uint256(keccak256(abi.encode("optimism.deployconfig"))))));
+    /// @notice 10_000 ether
+    bytes internal constant DEV_ACCOUNT_FUND_AMT = hex'021E19E0C9BAB2400000';
+
+    string internal genesisPath;
+    L2GenesisFixtures l2GenesisFixtures;
+
+    /// @notice Default Anvil dev accounts. Only funded if `cfg.fundDevAccounts == true`.
+    address[10] internal devAccounts = [
+        0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, 0x70997970C51812dc3A010C7d01b50e0d17dc79C8,
+        0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC, 0x90F79bf6EB2c4f870365E785982E1f101E93b906,
+        0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65, 0x9965507D1a55bcC2695C58ba16FB37d819B0A4dc,
+        0x976EA74026E726554dB657fA54763abd0C3a0aa9, 0x14dC79964da2C08b23698B3D3cc7Ca32193d9955,
+        0x23618e81E3f5cdF7f54C3d65f7FBc0aBf5B21E8f, 0xa0Ee7A142d267C1f36714E4a8F75612F20a79720
+    ];
+
+    struct StorageData {
+        bytes32 key;
+        bytes32 value;
+    }
+
+    /// @notice `balance` and `nonce` are being parsed as `bytes` even though their JSON representations are hex strings.
+    ///         This is because Foundry has a limitation around parsing strings as numbers when using `vm.parseJson`,
+    ///         and because we're using `abi.decode` to convert the JSON string, we can't use coersion (i.e. `vm.parseJsonUint`)
+    ///         to tell Foundry that the strings are numbers. So instead we treat them as `byte` strings and parse as
+    ///         `uint`s when needed. Additional context: https://github.com/foundry-rs/foundry/issues/3754
+    struct Alloc {
+        address addr;
+        bytes balance;
+        bytes code;
+        bytes nonce;
+        StorageData[] storageData;
+    }
+
+    /// @notice Checks if deploy config file exists at `deployConfigPath`, if it doesn't
+    ///         reverts with warning to run `make devnet-allocs` at monorepo root.
+    ///         Afterwards, checks if genesis file exists at `genesisPath`, and if not
+    ///         runs `L2Genesis.s.sol` script to generate it.
+    function setUp() public override {
+        super.setUp();
+        Artifacts.setUp();
+
+        string memory deployConfigPath = string.concat(vm.projectRoot(), "/deploy-config/", deploymentContext, ".json");
+        if (!vm.exists(deployConfigPath)) {
+            revert(string.concat(
+                "Did not find deploy config at: ",
+                deployConfigPath,
+                ", try running make devnet-allocs in monorepo root"
+            ));
+        }
+
+        vm.etch(address(cfg), vm.getDeployedCode("DeployConfig.s.sol:DeployConfig"));
+        vm.label(address(cfg), "DeployConfig");
+        vm.allowCheatcodes(address(cfg));
+        cfg.read(deployConfigPath);
+
+        genesisPath = string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/genesis-l2.json");
+        if (!vm.exists(genesisPath)) {
```

Comment:
You can import the contract and execute it directly, no need to ffi into cli to call foundry

Reply:
I refactored to the following:

```solidity
genesisPath = string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/genesis-l2.json");
if (!vm.exists(genesisPath)) {
    L2Genesis l2Genesis = new L2Genesis();
    l2Genesis.setUp();
    l2Genesis.run();
}
```

and while it does execute the script, it seems to have an issue reading the genesis outfile immediately after it writes it (maybe a race condition):

```
Traces:
  [2384521531] L2Genesis_Test::test_allocs()
    ├─ [0] VM::ffi(["bash", "-c", "jq -cr 'to_entries | map({addr: .key, balance: .value.balance, code: .value.code, nonce: .value.nonce, storageData: (.value.storage | to_entries | map({key: .key, value: .value}))})' /Users/spacesailor/Public/code/optimism/packages/contracts-bedrock/deployments/devnetL1/genesis-l2.json"])
    │   └─ ← 0x...
    ├─ [0] VM::parseJson("<JSON file>") [staticcall]
    │   └─ ← <encoded JSON value>
    ├─ [0] DeployConfig::getDevAccounts() [staticcall]
    │   └─ ← ()
    └─ ← EvmError: Revert

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 1.10s

Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/scripts/L2Genesis.t.sol:L2Genesis_Test
[FAIL. Reason: EvmError: Revert] test_allocs() (gas: 2384521531)
```

Running the test again after the above error yields a revert because of an unexpected alloc (I'm not sure what it is, probably the deploy config or a forge util contract?):

```json
"0x7fa9385be102ac3eac297483dd6233d62b3e1496": {
    "balance": "0xffffffffffffffffffffffff",
    "code": "0x...",
    "nonce": "0x2",
    "storage": {
      "0x000000000000000000000000000000000000000000000000000000000000001e": "0x00000000000000000000000000000000000000000000000000000000000000af",
      "0x000000000000000000000000000000000000000000000000000000000000001f": "0x00000000000000000000000000000000000000000000000000000000000000bf",
      "0x0000000000000000000000000000000000000000000000000000000000000020": "0x6465766e65744c31000000000000000000000000000000000000000000000010",
      "0x0000000000000000000000000000000000000000000000000000000000000021": "0x00000000000000000000000000000000000000000000000000000000000000cf",
      "0x3a6357012c1a3ae0a17d304c9920310382d968ebcc4b1771f41c6b304205b570": "0x2f55736572732f73706163657361696c6f722f5075626c69632f636f64652f6f",
      "0x3a6357012c1a3ae0a17d304c9920310382d968ebcc4b1771f41c6b304205b571": "0x7074696d69736d2f7061636b616765732f636f6e7472616374732d626564726f",
      "0x3a6357012c1a3ae0a17d304c9920310382d968ebcc4b1771f41c6b304205b572": "0x636b2f6465706c6f796d656e74732f6465766e65744c312f67656e657369732d",
      "0x3a6357012c1a3ae0a17d304c9920310382d968ebcc4b1771f41c6b304205b573": "0x6c322e6a736f6e00000000000000000000000000000000000000000000000000",
      "0x50bb669a95c7b50b7e8a6f09454034b2b14cf2b85c730dca9a539ca82cb6e350": "0x2f55736572732f73706163657361696c6f722f5075626c69632f636f64652f6f",
      "0x50bb669a95c7b50b7e8a6f09454034b2b14cf2b85c730dca9a539ca82cb6e351": "0x7074696d69736d2f7061636b616765732f636f6e7472616374732d626564726f",
      "0x50bb669a95c7b50b7e8a6f09454034b2b14cf2b85c730dca9a539ca82cb6e352": "0x636b2f6465706c6f796d656e74732f6465766e65744c31000000000000000000",
      "0xa03837a25210ee280c2113ff4b77ca23440b19d4866cca721c801278fd08d807": "0x2f55736572732f73706163657361696c6f722f5075626c69632f636f64652f6f",
      "0xa03837a25210ee280c2113ff4b77ca23440b19d4866cca721c801278fd08d808": "0x7074696d69736d2f7061636b616765732f636f6e7472616374732d626564726f",
      "0xa03837a25210ee280c2113ff4b77ca23440b19d4866cca721c801278fd08d809": "0x636b2f6465706c6f796d656e74732f6465766e65744c312f2e6465706c6f7900"
    }
  }
```

Reply:
We shouldn't ffi or call run in this script. `setUp` should ensure that all config is read properly so that we can call functions inside of `run` within individual tests

Reply:
I wonder if you do `etch(address(L2Genesis), "");` before dumping the state it will help to ensure that the contract doesn't exist. Another way to prevent this is to make the test contract inherit `L2Genesis` because the test contract is filtered out automatically